### PR TITLE
[release/2.7] github.com/golang-jwt/jwt v3.2.2

### PR DESCRIPTION
**Pull request by thaJeztah**
relates to https://github.com/distribution/distribution/issues/3361
relates to https://github.com/distribution/distribution/pull/3459 (main branch)
alternative to https://github.com/distribution/distribution/pull/3465
closes https://github.com/distribution/distribution/pull/3465

~Using the https://github.com/form3tech-oss/jwt-go.git, which is actively maintained.~

~full diff: https://github.com/form3tech-oss/jwt-go/compare/a601269ab70c205d26370c16f7c81e9017c14e04...v3.2.4~

to address CVE-2020-26160

full diff: https://github.com/golang-jwt/jwt/compare/a601269ab70c205d26370c16f7c81e9017c14e04...v3.2.2

3.2.1 release notes
---------------------------------------

- Import Path Change: See MIGRATION_GUIDE.md for tips on updating your code
  Changed the import path from github.com/dgrijalva/jwt-go to github.com/golang-jwt/jwt
- Fixed type confusion issue between string and []string in VerifyAudience.
  This fixes CVE-2020-26160

3.2.2 release notes
---------------------------------------

- Starting from this release, we are adopting the policy to support the most 2
  recent versions of Go currently available. By the time of this release, this
  is Go 1.15 and 1.16.
- Fixed a potential issue that could occur when the verification of exp, iat
  or nbf was not required and contained invalid contents, i.e. non-numeric/date.
  Thanks for @thaJeztah for making us aware of that and @giorgos-f3 for originally
  reporting it to the formtech fork.
- Added support for EdDSA / ED25519.
- Optimized allocations.

### CODE REVIEW

**Comment by thaJeztah** [comment]
From https://github.com/distribution/distribution/pull/3465#discussion_r679129891

> Not against using a replace if vndr can do that. Should we consider using the fork which the original repository now links to? https://github.com/golang-jwt/jwt

Looks like there's another fork as well (and that's the "official" replacement?)

File: vendor/github.com/dgrijalva/jwt-go/map_claims.go [hunk 1]
```diff
@@ -25,49 +27,37 @@ func (m MapClaims) VerifyAudience(cmp string, req bool) bool {
...
 
 // Compares the exp claim against cmp.
 // If required is false, this method will return true if the value matches or is unset
 func (m MapClaims) VerifyExpiresAt(cmp int64, req bool) bool {
-	exp, ok := m["exp"]
-	if !ok {
-		return !req
-	}
-	switch expType := exp.(type) {
+	switch exp := m["exp"].(type) {
```
Comment (thaJeztah) [inline vendor/github.com/dgrijalva/jwt-go/map_claims.go]: Looks like the official repository does not have the security fix that the `orm3tech-oss/jwt-go` fork has; switching to `github.com/golang-jwt/jwt` removes these changes, which seem to be related to that security fix: https://github.com/form3tech-oss/jwt-go/compare/v3.2.3...v3.2.4

Comment (brackendawson) [reply vendor/github.com/dgrijalva/jwt-go/map_claims.go]: I believe golang-jwt/jwt v3.2.1 fixed CVE-2020-26160 with different code: https://github.com/golang-jwt/jwt/commit/0f726ea0e7256f9056da372f822bd534f9d1dc5f

Comment (thaJeztah) [reply vendor/github.com/dgrijalva/jwt-go/map_claims.go]: The equivalent to that one is in https://github.com/form3tech-oss/jwt-go/commit/bb5e6d8f35bea762d2564725ef646dd6e6a57799 for the form3tech repository. Looks like v3.2.4 is fixing a different security issue?

Comment (thaJeztah) [reply vendor/github.com/dgrijalva/jwt-go/map_claims.go]: Correct. I suspect no CVE has been requested for the security fix in  form3tech-oss/jwt-go v3.2.4. If it's verified to be a security issue, a CVE should probably be requested (both for form3tech-oss/jwt-go, and for  github.com/golang-jwt/jwt / dgrijalva/jwt-go)

Comment (brackendawson) [reply vendor/github.com/dgrijalva/jwt-go/map_claims.go]: Shall we merge those two PRs while we wait for the situation with this new vulerability to be resolved upstream or do you think the workaround in this PR is a better fix for right now?

Comment (thaJeztah) [reply vendor/github.com/dgrijalva/jwt-go/map_claims.go]: My concern with https://github.com/distribution/distribution/pull/3465 is that it's pulling in 5.6k lines of code that's unrelated to the security fix (and which would have to be reviewed), with the goal to pull in a security fix that's < 50 lines. And (worse?) switching to a non-official fork (which, based on the `go.mod` addition, is at least somewhat "dubious" on quality).

From that perspective, I'd rather take the official fork (_ideally_ with https://github.com/golang-jwt/jwt/pull/40, if that's confirmed to be an issue), as is done in this PR.

For the `main` branch, I don't think there's an direct urgency to get it merged, unless a `v3.0.0` release is pending.

**Comment by thaJeztah** [comment]
Looks like https://github.com/golang-jwt/jwt is the best option, and https://github.com/golang-jwt/jwt/pull/40 is currently being reviewed, so I squashed the commits, and went for that fork (as it's the official replacement).

**Comment by thaJeztah** [comment]
Updated to v3.2.2, which was just released

**Comment by thaJeztah** [comment]
Looks like we need to update Golang as well;

```
/docker/distribution/vendor/github.com/dgrijalva/jwt-go/ecdsa.go:135:5: r.FillBytes undefined (type *big.Int has no field or method FillBytes
```

**Comment by thaJeztah** [comment]
ok, works with Go 1.16, but should be moved to a separate PR

**Comment by thaJeztah** [comment]
opened https://github.com/distribution/distribution/pull/3472

**Comment by thaJeztah** [comment]
Rebased, because https://github.com/distribution/distribution/pull/3472 was merged

**Comment by thaJeztah** [comment]
Thinking a bit more about this PR: does this project actually perform the validation of the JWT token? Because if not, then this patch may not be needed at all

**Comment by milosgajdos** [comment]
@thaJeztah distribution has `token` auth package that does verification https://github.com/distribution/distribution/blob/release/2.7/registry/auth/token/token.go

**Review by milosgajdos — APPROVED** [review]

**Comment by thaJeztah** [comment]
> @thaJeztah distribution has token auth package that does verification

Ah, thanks for checking!

**Review by wy65701436 — APPROVED** [review]
lgtm

**Pull request merged**